target_bigquery.py

```diff
@@ -265,57 +276,276 @@ def persist_lines_stream(project_id, dataset_id, lines=None, validate_records=Tr
+            if rows.get(stream):
+                write_rows_to_bigquery([stream], emit_state_after_write=True)
+
+            # If stream in `bookmarks` doesn't have `replication_key_value` we assume this state is
+            # a first one for a particular stream and recreate table.
+            # See: https://github.com/singer-io/tap-mysql#incremental
+            rep_key = state.get("bookmarks", {}).get(full_stream, {}).get("replication_key_value")
+            # NOTE: this will only work if `SchemaMessage` already received before
+            if (
+                stream
+                and not rep_key
+                and not tables[stream].schema == build_schema(schemas[stream])
+            ):
+                # Delete table
+                # TODO: instead of deleting table maybe we could write to a temporary table and then
+                # replace after finishing parsing the last line?
+                table_ref = f"{dataset_ref}.{stream}"
+                logger.info(f"Deleting table: {table_ref}")
+                bigquery_client.delete_table(table_ref)
+
+                # Recreate table
+                tables[stream] = bigquery_client.create_table(
+                    bigquery.Table(table_ref, schema=build_schema(schemas[stream]))
+                )
+                logger.info(
+                    f"Created table '{tables[stream]}' with schema: {tables[stream].schema}"
+                )
+                # Mark the recreated table so we know we need to retry if first attempts fail
+                recreated_tables[stream] = True
+
+        elif isinstance(msg, singer.SchemaMessage):
+            stream = msg.stream
+            schemas[stream] = msg.schema
+            key_properties[stream] = msg.key_properties
+            table_ref = f"{dataset_ref}.{stream}"
+            try:
+                tables[stream] = bigquery_client.get_table(table_ref)
+            except api_core.exceptions.NotFound:
+                # This will happen on the very first run
+                tables[stream] = bigquery_client.create_table(
+                    bigquery.Table(table_ref, schema=build_schema(schemas[stream]))
+                )
+            rows[stream] = []
+            errors[stream] = []
+
+        elif isinstance(msg, singer.ActivateVersionMessage):
+            # This is experimental and won't be used yet
+            pass
+
+        else:
+            raise Exception(f"Unrecognized message: {msg}")
+            failed_lines.append(msg)
+
+    # We shouldn't have any rows left to write, but let's try just in case
+    write_rows_to_bigquery(rows.keys())
+
+    if failed_lines:
+        logger.warning(f"Number of failed lines: {len(failed_lines)}")
+        # TODO: maybe also write failed_lines to special table in BigQuery
+        try:
+            # NOTE: if we have too many failed lines we might not be able to write the log entry
+            logger.warning(f"Failed lines: {str(failed_lines)}")
```

Comment:
Could write each line separately so we don't risk generating an exception while writing?

Reply:
Good that you mention, I actually had an idea over the weekend that the best way would probably be writing just the ids and then even if thousands of lines failed it would still work.

Reply:
Hmm, actually it's a bit trickier than I thought because we might have all different types of failed lines in here, so I might leave this unaddressed for now and do in another PR...

Reply:
Why not log them out as they happen? Do you want to keep all the failures together?

Reply:
I was reluctant to spam the logs in case a whole 5000-row chunk fails.

Also, there is a plan to write these into a special table in BQ for possible recovery: https://app.clubhouse.io/beryl/story/6868/write-rows-which-we-can-t-insert-into-its-proper-bigquery-destination-table-into-a-dead-letter-table